### PR TITLE
Use str.isidentifier to match idents on python 3

### PR DESCRIPTION
**Pull request by mitsuhiko**
Not sure I like this very much and I need to ensure still that `\w` is a superset of the identifier ranges. It is however also what tokenize does on python 3 so there is precedent. Moreover though the error message is different now for bad identifiers but I assume not many users will notice.

Refs #729

### CODE REVIEW

File: jinja2/lexer.py [hunk 1]
```diff
@@ -577,6 +565,10 @@ def wrap(self, stream, name=None, filename=None):
                 token = value
             elif token == 'name':
                 value = str(value)
+                if check_ident and not value.isidentifier():
```
Comment (lf-) [inline jinja2/lexer.py]: This isn't sufficient guarding against python2 for some reason. It still raises an AttributeError.

File: jinja2/lexer.py [hunk 2]
```diff
@@ -34,28 +34,16 @@
...
-
-    return name_re
-
 # we use the unicode identifier rule if this python version is able
 # to handle unicode identifiers, otherwise the standard ASCII one.
-name_re = _make_name_re()
-del _make_name_re
+try:
+    compile('föö', '<unknown>', 'eval')
+except SyntaxError:
```
Comment (lf-) [inline jinja2/lexer.py]: Bug! Python3 doesn't raise SyntaxError on unicode identifiers. There's a logical inversion here.

**Comment by lf-** [comment]
I confirmed that `\w` matches all meaningful identifiers:

```python
import re
import sys
import unicodedata

cre = re.compile(r'\w')
for cp in range(sys.maxunicode + 1):
    s = chr(cp)
    if s.isidentifier() and not re.match(cre, s):
        print(hex(cp), unicodedata.name(s))
```

Valid identifiers not matched by `\w`:

```
import re
import sys
import unicodedata

cre = re.compile(r'\w')
for cp in range(sys.maxunicode + 1):
    s = chr(cp)
    if s.isidentifier() and not re.match(cre, s):
        print(hex(cp), unicodedata.name(s))
```

Beyond that, it appears that the clauses for the try-except at the start are backwards, causing python 2 to raise and python 3 to not have unicode identifier support.

**Comment by mitsuhiko** [comment]
Need to add a test to this one but I fixed the reverse thing and added characters missing. Oddly enough I can only reproduce `\w` not matching on the latter two but not the mongolian characters.

**Comment by davidism** [comment]
Need to also test `not keyword.iskeyword(value)` to catch things like `True` and `class` which `str.isidentifier` returns true for on some Python versions.

**Comment by davidism** [comment]
`env.from_string('{{ ℮ }}')` (or any of the four special case characters) raises `TemplateSyntaxError: unexpected char '℮' at 3`. Other Unicode characters work as expected.

**Comment by davidism** [comment]
Pushed a simple test, currently fails for the four special case characters.

**Comment by davidism** [comment]
Weird, all four fail on my machine, but only the two Mongolian characters fail on Travis.

**Comment by davidism** [comment]
The regex is specifically failing on the `\b` boundaries, if they are removed the special cases work. This is because `\b` means "transition from `\w` to `\W`, and the special cases are not in `\w`.

**Comment by mitsuhiko** [comment]
I wonder if that was not already broken before.

**Comment by davidism** [comment]
Just checked, `e.from_string('{{ ℮ }}')` works on master.

**Comment by davidism** [comment]
I don't think the `\b` is necessary, it's not present in the previous stringdef regex. All tests still pass with the boundaries removed.

**Comment by davidism** [comment]
I reported the `\w` issue to Python: http://bugs.python.org/issue30838

@lf- thanks for helping figure this out.

**Comment by davidism** [comment]
Writing more tests based on the previous stringdef, https://raw.githubusercontent.com/pallets/jinja/2.9.5/jinja2/_stringdefs.py, turns up some invalid characters in the previous regex. The following start characters were matched but were not valid as start or continue characters:

```
\u309B KATAKANA-HIRAGANA VOICED SOUND MARK
\u309C KATAKANA-HIRAGANA SEMI-VOICED SOUND MARK
```

The following were matched and are valid continue characters, but are no longer matched by `\w`:

```
· MIDDLE DOT
· GREEK ANO TELEIA
```

**Comment by davidism** [comment]
Then I realized we were missing a test in that script: `('a' + c).isidentifier()` for characters that are valid for continue but not start. That reveals a lot more that aren't matched by `\w`.

**Comment by davidism** [comment]
Going to have to bring back `_stringdefs`, but still simplified.

**Comment by davidism** [comment]
Original `_stringdefs`:

```
         len   sizeof
start    48194 96462
continue 49826 99726
total    98020 196188
```

New `_identifier`:

```
len: 635
sizeof: 2616
```

Timing is hard to test, but on my machine `re.purge(); re.compile(...)` is down from ~100 ms to ~840 µs.

Strategy was to collapse contiguous ranges and rely on `str.isidentifier` to validate so that the regex is simpler.

**Comment by mitsuhiko** [comment]
That sounds good. Ideally we can reuse the same thing as we had before to regenerate the regex though we need to make sure we hit the set that is missing in all versions of Python 3.

**Comment by davidism** [comment]
At the cost of about twice as much space, the regex could be made more accurate by omitting `\w` and generating the full range of valid characters. At the cost of twice again as much space, we could remove the need for calling `str.isidentifer` during lexing by matching start and continue characters in the regex. For now, I'm happy with where we are and will leave `\w` and `isidentifier` in.

**Pull request merged**